File: dawn_cpp/rmq.cpp
#include <algorithm>
#include <cmath>
#include <vector>

struct Tree {
    std::vector<int> values;
    std::vector<int> euler;
    std::vector<int> first;
    std::vector<std::vector<int>> rmq;
};

int log2(int n) {
    return static_cast<int>(std::ceil(std::log2(n)));
}

void precomputeRMQ(Tree& tree) {
    int n = tree.euler.size();
    int k = log2(n) + 1;
    tree.rmq = std::vector<std::vector<int>>(k, std::vector<int>(n));

    for (int i = 0; i < n; ++i) {
        tree.rmq[0][i] = i;
    }

    for (int j = 1; j < k; ++j) {
        for (int i = 0; i + (1 << (j - 1)) < n; ++i) {
            int x = tree.rmq[j - 1][i];
            int y = tree.rmq[j - 1][i + (1 << (j - 1))];
            tree.rmq[j][i] = (tree.euler[x] < tree.euler[y]) ? x : y;
        }
    }
}

int query(const Tree& tree, int l, int r) {
    l = std::min(l, r);
    r = std::max(l, r);
    int k = log2(r - l + 1);
    int x = tree.rmq[k][tree.first[l]];
    int y = tree.rmq[k][tree.first[r] - (1 << k) + 1];
    return std::min(tree.values[tree.euler[x]], tree.values[tree.euler[y]]);
}

Tree fromTree(const std::vector<int>& values) {
    int n = values.size();
    std::vector<int> euler(2 * n);
    std::vector<int> first(2 * n);

    auto buildEuler = [&](int tree, int i) {
        int j = i + 1;
        euler[i] = tree;
        first[i] = j;
        if (j < 2 * n) {
            buildEuler(tree + 1, j);
        }
    };
    buildEuler(0, 0);

    Tree tree = {values, euler, first, {}};
    precomputeRMQ(tree);
    return tree;
}

File: dawn_rust/src/rmq.rs
pub mod rmq {
    use std::cmp::{max, min};

    pub struct Tree<T> {
        values: Vec<T>,
        euler: Vec<usize>,
        first: Vec<usize>,
        rmq: Vec<Vec<usize>>,
    }

    impl<T: Ord + Copy> Tree<T> {
        fn log2(n: usize) -> usize {
            (n as f64).log2() as usize + 1
        }

        fn precompute_rmq(&mut self) {
            let n = self.euler.len();
            let k = Self::log2(n);
            self.rmq = vec![vec![0; n]; k];

            for i in 0..n {
                self.rmq[0][i] = i;
            }

            for j in 1..k {
                for i in 0..n - (1 << (j - 1)) {
                    let x = self.rmq[j - 1][i];
                    let y = self.rmq[j - 1][i + (1 << (j - 1))];
                    self.rmq[j][i] = if self.euler[x] < self.euler[y] {
                        x
                    } else {
                        y
                    };
                }
            }
        }

        pub fn query(&self, l: usize, r: usize) -> T {
            let l = min(l, r);
            let r = max(l, r);
            let k = Self::log2(r - l + 1);
            let x = self.rmq[k][self.first[l]];
            let y = self.rmq[k][self.first[r] - (1 << k) + 1];
            min(self.values[self.euler[x]], self.values[self.euler[y]])
        }

        pub fn from_tree(values: &[T]) -> Self
        where
            T: Ord + Copy,
        {
            let n = values.len();
            let mut euler = vec![0; 2 * n];
            let mut first = vec![0; 2 * n];

            let mut build_euler = |tree: usize, i: usize| {
                let j = i + 1;
                euler[i] = tree;
                first[i] = j;
                if j < 2 * n {
                    build_euler(tree + 1, j);
                }
            };
            build_euler(0, 0);

            let mut tree = Tree {
                values: values.to_vec(),
                euler,
                first,
                rmq: vec![],
            };
            tree.precompute_rmq();
            tree
        }
    }
}